### PR TITLE
Azure REST GET should respect nextLink when list resources

### LINKED ISSUE #603: Updating instance gives PrivateIPAddressIsAllocated in v35.5.0
### Versions (*please add them if you have more releases for this issue*)
CPI      Version: 35.5.0
BOSH     Version: 268.3.0
CF       Version: -
Diego    Version: -

### Stemcell
Version: bosh-azure-hyperv-ubuntu-trusty-go_agent  3586.54
Distro:
  - [x] Ubuntu
  - [ ] CentOS

### Environment
- Azure Environment:
  - [x] Azure
  - [ ] AzureChinaCloud
  - [ ] AzureUSGovernment
  - [ ] AzureGermanCloud
  - [ ] AzureStack
    - Version: (e.g. 1712 update or build number 180106.1)
  - [ ] Others:
- Location: (e.g. East US)
- Subscription ID:
- Resource Group Name:
- [ ] Use Managed Disks

### Issue Details

When redeploying an existing deployment, forcing the CPI to recreate the instance by changing the vm_type, we get the error below.
With version 35.4.0 the workflow described below succeeds.

### Logs (*Please redact all sensitive information in your logs*)
Error message: {
  "error": {
    "code": "PrivateIPAddressIsAllocated",
    "message": "IP configuration /subscriptions/<...>/resourceGroups/rg-azure-staging/providers/Microsoft.Network/networkInterfaces/<...>/ipConfigurations/ipconfig0 is using the private IP address 10.0.0.1 which is already allocated to resource /subscriptions/<...>/resourceGroups/rg-azure-staging/providers/Microsoft.Network/networkInterfaces/<...>.",
    "details": []
  }
}>

### Repro steps (*if necessary, delete otherwise*)
* deploy [dummy](https://github.com/pivotal-cf-experimental/dummy-boshrelease)

```
---
name: dummy

releases:
- name: dummy
  version: latest

stemcells:
- alias: ubuntu
  os: ubuntu-trusty
  version: latest

instance_groups:
- name: dummy
  instances: 1
  vm_type: vm_2cpu_8gb
  stemcell: ubuntu
  networks:
  - name: default
    static_ips:
      - 10.0.0.1
  azs:
  - z1
  jobs:
  - name: dummy
    release: dummy

update:
  canaries: 1
  max_in_flight: 6
  serial: false
  canary_watch_time: 1000-60000
  update_watch_time: 1000-60000
```

* change vm_type and redeploy

```
---
name: dummy

releases:
- name: dummy
  version: latest

stemcells:
- alias: ubuntu
  os: ubuntu-trusty
  version: latest

instance_groups:
- name: dummy
  instances: 1
  vm_type: vm_4cpu_8gb
  stemcell: ubuntu
  networks:
  - name: default
    static_ips:
      - 10.0.0.1
  azs:
  - z1
  jobs:
  - name: dummy
    release: dummy

update:
  canaries: 1
  max_in_flight: 6
  serial: false
  canary_watch_time: 1000-60000
  update_watch_time: 1000-60000
```

### PR DESCRIPTION
**Pull request by gossion**
To fix #603

### CODE REVIEW

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 1]
```diff
@@ -123,10 +123,44 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
...
+    # get single resource
+    # example: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkInterfaces/{networkInterfaceName}
     def get_resource_by_id(url, params = {})
       result = nil
       begin
-        result = http_get(url, params)
+        uri = http_url(url, params)
+        response = http_get(uri)
+        result = JSON.parse(response.body) unless response.body.nil?
+      rescue AzureNotFoundError => e
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: log it down?

File: src/bosh_azure_cpi/Gemfile [hunk 2]
```diff
@@ -7,11 +7,13 @@ gem 'bosh_common',      '1.3262.24.0'
...
 gem 'jwt',              '1.5.6'
+gem 'deep_merge',       '~> 1.0', '>= 1.0.1'
 
 group :development, :test do
   gem 'rake',        '~>12.0'
   gem 'rspec',       '~>3.5.0'
   gem 'rubocop',     '~>0.60.0', require: false
   gem 'simplecov'
   gem 'webmock'
+  gem 'deep_merge',  '~> 1.0', '>= 1.0.1'
```
Comment (bingosummer) [inline src/bosh_azure_cpi/Gemfile]: duplicate?

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 3]
```diff
@@ -123,10 +123,44 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
...
+        uri = http_url(url, params)
+        response = http_get(uri)
+        unless response.body.nil?
+          body = JSON.parse(response.body)
+          result = body
+          next_url = body['nextLink']
+        end
+
+	while next_url != nil
+          @logger.warn("Getting resources from nextLink #{next_url}")
```
Comment (bingosummer) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: why is it a warning message?

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 4]
```diff
@@ -123,10 +123,44 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
...
+          uri = URI(next_url)
+          response = http_get(uri)
+          cloud_error("Got empyt page from nextLink #{next_url}") if response.body.nil?
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: typo: empty.

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 5]
```diff
@@ -123,10 +123,44 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
...
+      begin
+        uri = http_url(url, params)
+        response = http_get(uri)
+        unless response.body.nil?
+          body = JSON.parse(response.body)
+          result = body
+          next_url = body['nextLink']
+        end
+
+	while next_url != nil
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: minor: URI seems may throws the BadURIError. we may better to catch it too though we should trust azure api more.

Comment (bingosummer) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: just out of curious, is it possble that next_url is an empty string?

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 6]
```diff
@@ -123,11 +123,47 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
...
       rescue AzureNotFoundError => e
+        @logger.debug("Resource not found for url #{url} with parms #{params}")
+        result = nil
+      end
+      result
+    end
+
+    # get list of resources
+    # example: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkInterfaces
+    def get_resources_by_url(url, params = {})
```
Comment (bingosummer) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: do we need to add unit test in this PR?

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb (hunk 5 above)
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: I think Azure should response a valid URL, if it isn't a valid URL, it is reasonable for URI to raise a BadURIError.

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb (hunk 3 above)
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: typo. should be 'debug'

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb (hunk 6 above)
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: yes, we should. but busy on other things. however, this is already tested manually.

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb (hunk 5 above)
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: not sure, did not see that case.

File: src/bosh_azure_cpi/Gemfile (hunk 2 above)
Comment (gossion) [reply src/bosh_azure_cpi/Gemfile]: you are right. fixed

**Pull request merged**